### PR TITLE
new: cache noise used in the end to reduce computation time

### PR DESCRIPTION
**Pull request by jaskarth**
This PR speeds up end noise generation by a whopping 200% on my machine by caching the noise used to calculate biomes and island density. Vanilla samples a 25x25 region of noise for these calculations, which add up very quickly as it doesn't store any of the values for later reuse. By adopting the same type of cache used in the fast layer sampler, we can use many of the previous values from old calculations and avoid recalculating the noise. I've attached some rough profiler measurements while generating terrain the end.
Before the optimization:
image
Afterwards:
image
By caching the noise, we are spending 4x less time on the noise calculation. Optimizing this more isn't possible due to the modulo operator within the code.
This change is fully compatible with vanilla generation.

Thanks to @gegy1000 for writing most of the code here 😄

### CODE REVIEW

File: src/main/java/me/jellysquid/mods/lithium/mixin/world/fast_island_noise/MixinSurfaceChunkGenerator.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.Redirect;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+import net.minecraft.util.math.noise.SimplexNoiseSampler;
+import net.minecraft.world.biome.source.BiomeSource;
+import net.minecraft.world.gen.chunk.ChunkGeneratorType;
+import net.minecraft.world.gen.chunk.SurfaceChunkGenerator;
+
+@Mixin(SurfaceChunkGenerator.class)
+public class MixinSurfaceChunkGenerator {
```
Comment (jellysquid3) [inline src/main/java/me/jellysquid/mods/lithium/mixin/world/fast_island_noise/MixinSurfaceChunkGenerator.java:17]: This PR claims that it speeds up End generation, but isn't this also implementing the noise cache into the overworld chunk generator?

**Review by jellysquid3 — APPROVED** [review]
LGTM.

**Comment by jaskarth** [comment]
That SurfaceChunkGenerator patch is what the end uses to calculate the density of the surrounding terrain. It's only used when `useIslandOverrideNoise` is enabled in the chunk generator config, which is only true for the end dimension's chunk generator and chunk generators added via datapacks.

File: src/main/java/me/jellysquid/mods/lithium/common/world/noise/SimplexNoiseCache.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                    float deltaX = gridOriginX - offsetX * GRID_SIZE;
+                    float deltaZ = gridOriginZ - offsetZ * GRID_SIZE;
+
+                    // Calculate the density at this grid cell
+                    float scaledDistance = MathHelper.sqrt(deltaX * deltaX + deltaZ * deltaZ) * distanceFactor;
+                    float densityHere = ISLAND_RADIUS - scaledDistance;
+
+                    // Try to return early if we're over the max
+                    if (densityHere > density) {
+                        if (density >= MAX) {
```
Comment (Kroppeb) [inline src/main/java/me/jellysquid/mods/lithium/common/world/noise/SimplexNoiseCache.java]: Shouldn't this be `densityHere >= MAX`

Comment (jaskarth) [reply src/main/java/me/jellysquid/mods/lithium/common/world/noise/SimplexNoiseCache.java]: Yeah, good catch! It shouldn't have caused any problems, but that shaves off another grid cycle 😅

File: src/main/java/me/jellysquid/mods/lithium/mixin/world/fast_island_noise/MixinSurfaceChunkGenerator.java (hunk 1 above)
Comment (Gegy) [reply src/main/java/me/jellysquid/mods/lithium/mixin/world/fast_island_noise/MixinSurfaceChunkGenerator.java:17]: The "surface" (??) chunk generator now functions both for the overworld and the end. I think it was a pretty recent change though (oh lol scoder replied already ignore me)

**Review by Titaniumtown — APPROVED** [review]
LGTM, when will this be merged?

**Comment by 2No2Name** [comment]
Will test the changes now. Going to merge if everything goes well

**Comment by 2No2Name** [comment]
LGTM. Rebasing onto 1.1.6.x/dev

**Comment by Titaniumtown** [comment]
*1.16.x/dev

**Pull request closed without merging**